airflow/jobs/local_task_job.py

```diff
@@ -135,7 +135,11 @@ def heartbeat_callback(self, session=None):
             self.task_runner.terminate()
             return
 
-        self.task_instance.refresh_from_db()
+        if session is None:
+            self.task_instance.refresh_from_db()
+        else:
+            self.task_instance.refresh_from_db(session=session)
```

Comment:
Not sure about this one. I think it would be better to validate the session in `provide_session` WDYT?

Reply:
Great point. I have thought about it. The concern is that if we enforce it in the `provide_session`, then there is no way to  pass in `session=None` to the function.

Reply:
```python
def provide_session(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        arg_session = 'session'

        func_params = func.__code__.co_varnames
        session_in_args = arg_session in func_params and \
            func_params.index(arg_session) < len(args)
        session_in_kwargs = arg_session in kwargs

       # Here we validate that the session is in kwargs/args and that it's value is
       # not None so we can pass if further, otherwise we create new session
        if session_in_kwargs and kwargs[arg_session] is not None:
            return func(*args, **kwargs)
        elif session_in_args and func_params[func_params.index(arg_session)] is not None:
            return func(*args, **kwargs)
        else:
            with create_session() as session:
                kwargs[arg_session] = session
                return func(*args, **kwargs)

    return wrapper

Reply:
yep, that was the place I was trying to change. If we always change `None` to a real `session` object, the users will not be able to set the `session` to `None`, (as the decorator always changes it)  when they don't want to create a session. It depends on whether we want `provide_session` to always create a `session` even when the user explicitly says i don't want a session to be created.

Reply:
Hm, but if user will provide `session=None` then `session.query(Model)` will raise an error, right?

Reply:
After more thoughts, it looks like we don't actually need this PR, as the airflow use `scoped_session`, which will use the same session, no matter whether we pass the `session` to the `refresh_from_db`